src/Framework/LazyFormattedBuildEventArgs.cs

```diff
@@ -125,20 +130,17 @@ public override string Message
         /// <param name="writer">Binary writer which is attached to the stream the event will be serialized into.</param>
         internal override void WriteToStream(BinaryWriter writer)
         {
-            // Locking is needed here as this is invoked on the serialization thread,
-            // whereas a local logger (a distributed logger) may concurrently invoke this.Message
-            // which will trigger formatting and thus the exception below
-            lock (locker)
+            string formattedMessage = base.Message;
+            if (formattedMessage != null)
             {
-                bool hasArguments = arguments != null;
-                base.WriteToStream(writer);
-
-                if (hasArguments && arguments == null)
-                {
-                    throw new InvalidOperationException("BuildEventArgs has formatted message while serializing!");
-                }
-
-                if (arguments != null)
+                // We've already formatted the message, write it to the stream and omit arguments.
+                base.WriteToStreamWithExplicitMessage(writer, formattedMessage);
+                writer.Write(-1);
+            }
+            else
+            {
+                base.WriteToStreamWithExplicitMessage(writer, unformattedMessage);
+                if (arguments?.Length > 0)
```

Comment:
Understanding check: There's no longer a race here because we no longer set `arguments = null` on formatting? Does that mean that we hold onto objects in `arguments` for a longer time? Is that bad?

Reply:
I believe that in vast majority of cases formatting is at the end of EventArgs litetime. It is taken out of queue, sinked to all loggers, some logger might format it, and than collected by next `GC.Collect`. In both, former and new behavior, both EventArgs and its `this.arguments` will be collected be same GC collection.